[TCSACR-481] Add Tizen.Uix.Sticker guide (#1711)
Samsung/tizen-docs: docs/application/dotnet/guides/uix/sticker.md

# Sticker

> The Sticker feature provides users with fun experiences to share sticker information. The Sticker applications are composed of a [Sticker provider application](#provider_application) and [Sticker consumer applications](#consumer_application). The Sticker provider creates stickers and you can create your own Sticker consumer applications that use the stickers created by the Sticker provider. The Sticker provider application can store sticker details such as URI, keyword, and group name in the Sticker database. The Sticker consumer applications can retrieve sticker information using the Sticker API.

Reviewer: ```suggestion
The Sticker feature provides users with fun experiences to share sticker information. The Sticker applications are composed of a [Sticker provider application](#provider_application) and [Sticker consumer applications](#consumer_application). The Sticker provider creates stickers, and you can create your own Sticker consumer applications that use the stickers created by the Sticker provider. The Sticker provider application can store sticker details such as URI, keyword, and group name in the Sticker database. The Sticker consumer applications can retrieve sticker information using the Sticker API.
```

Reviewer 2: Dear @omar-srbd,

Inhong who is the author is absent due to attending the training session, and I can't modify in his forked repository.
I'll upload the separate PR to apply your suggestion.
Would you please review first except this suggestion?

Reviewer: @imfine98, thanks for informing the matter.
Since this is a very minuscule change and the author is absent I will approve this PR and make the suggested change by myself in another PR in the near future.

Reviewer 2: @omar-srbd,
Thank you very much.

1. To use the `Tizen.Uix.Sticker.StickerConsumer` class, the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

    ```
    <privileges>
       <privilege>http://tizen.org/privilege/mediastorage</privilege>
    </privileges>
    ```

2.  To use the methods and properties of the Tizen.Uix.Sticker namespace, include it in your application:

    ```csharp
    using Tizen.Uix.Sticker;
    ```

3. Initialize the Sticker consumer with the `Initialize()` method of the `Tizen.Uix.Sticker.StickerConsumer` class:

    ```csharp
    void Initialize()
    {
        StickerConsumer.Initialize();
    }
    ```

4. When the Sticker consumer is no longer needed, deinitialize it:

    ```csharp
    void Deinitialize()
    {
        StickerConsumer.Deinitialize();
    }
    ```

<a name="provider_application"></a>
## Create Sticker provider application

You can provide the sticker information to applications that use stickers. If you store the stickers in the Sticker database, the Sticker consumer applications can search for them.

<a name="set_sticker_data"></a>

> ### Set Sticker information to StickerData class

Reviewer: ```suggestion
### Set sticker information to StickerData class
```

Author: done

    ```csharp
    void InsertStickerData(StickerData stickerData)
    {
        try
        {
            StickerProvider.InsertData(stickerData);
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

- To store multiple sticker information, use `InsertData(string jsonPath)`:

    ```csharp
    async Task InsertStickerData(string jsonPath)
    {
        try
        {
            await StickerProvider.InsertData(jsonPath);
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

- To store sticker information without using `InsertData(StickerData data)` and `InsertData(string jsonPath)`, add the following metadata to the `tizen-manifest.xml` file:

  ```
  <category name="http://tizen.org/category/sticker"/>
  <!-- Sets the JSON local path -->
  <metadata key="http://tizen.org/metadata/sticker" value="/res/json/sticker.json"/>
  ```

  If you set the path of the JSON file to the value in metadata, sticker information of the JSON file is automatically stored when the Sticker provider application is installed.

  > [!NOTE]

>   >

Reviewer: Please remove the extra line after the alert.
```suggestion
```

Author: done

# Sticker

> The Sticker feature provides users with fun experiences to share sticker information. The Sticker applications are composed of a [Sticker provider application](#provider_application) and [Sticker consumer applications](#consumer_application). The Sticker provider creates stickers and you can create your own Sticker consumer applications that use the stickers created by the Sticker provider. The Sticker provider application can store sticker details such as URI, keyword, and group name to the Sticker database. The Sticker consumer applications can retrieve sticker information using the Sticker API.

Reviewer: ```suggestion
The Sticker feature provides users with fun experiences to share sticker information. The Sticker applications are composed of a [Sticker provider application](#provider_application) and [Sticker consumer applications](#consumer_application). The Sticker provider creates stickers and you can create your own Sticker consumer applications that use the stickers created by the Sticker provider. The Sticker provider application can store sticker details such as URI, keyword, and group name in the Sticker database. The Sticker consumer applications can retrieve sticker information using the Sticker API.
```

Author: done

> The main features of the Tizen.Uix.Sticker namespace include:

Reviewer: ```suggestion
The main features of the Tizen.Uix.Sticker namespace includes the following:
```

Author: done

- Preparing the Sticker service for use

  You can connect the Sticker daemon to operate the Sticker service.

- Sticker data

  You can use the `Tizen.Uix.Sticker.StickerData` class to [store the sticker information](#set_sticker_data) and to [get the sticker information](#get_sticker_data) stored in it.

- [Inserting the sticker information](#insert_sticker)

  - You can store the sticker information using the `Tizen.Uix.Sticker.StickerProvider` class at application runtime, and [use the JSON file](#set_json) to store the sticker information at a time.
  - You can store sticker information automatically when the Sticker provider application is installed without using the `Tizen.Uix.Sticker` methods.

- Retrieving the sticker information

  You can [retrieve the sticker information](#retrieve_sticker), which includes URI, keyword, group name, and so on.

**Figure: Sticker Mechanism**

![Sticker Mechanism](./media/sticker.png)

## Prerequisites

> To enable your Sticker provider application to use the Sticker functionality:

Reviewer: ```suggestion
To enable your Sticker provider application to use the Sticker functionality, follow these steps:
```

Author: done

1. To use the `Tizen.Uix.Sticker.StickerProvider` class, the following category must be added to the `tizen-manifest.xml` file:

    ```
    <category name="http://tizen.org/category/sticker"/>
    ```

2. To use the methods and properties of the Tizen.Uix.Sticker namespace, include it in your application:

    ```csharp
    using Tizen.Uix.Sticker;
    ```

3. Initialize the Sticker provider with the `Initialize()` method of the `Tizen.Uix.Sticker.StickerProvider` class:

    ```csharp
    void Initialize()
    {
        StickerProvider.Initialize();
    }
    ```

4. When the Sticker provider is no longer needed, deinitialize it:

    ```csharp
    void Deinitialize()
    {
        StickerProvider.Deinitialize();
    }
    ```

> To enable your Sticker consumer application to use the Sticker functionality:

Reviewer: ```suggestion
To enable your Sticker consumer application to use the Sticker functionality, follow these steps:
```

Author: done

1. To use the `Tizen.Uix.Sticker.StickerConsumer` class, the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

    ```
    <privileges>
       <privilege>http://tizen.org/privilege/mediastorage</privilege>
    </privileges>
    ```

2.  To use the methods and properties of the Tizen.Uix.Sticker namespace, include it in your application:

    ```csharp
    using Tizen.Uix.Sticker;
    ```

3. Initialize the Sticker consumer with the `Initialize()` method of the `Tizen.Uix.Sticker.StickerConsumer` class:

    ```csharp
    void Initialize()
    {
        StickerConsumer.Initialize();
    }
    ```

4. When the Sticker consumer is no longer needed, deinitialize it:

    ```csharp
    void Deinitialize()
    {
        StickerConsumer.Deinitialize();
    }
    ```

<a name="provider_application"></a>

> ## Create Sticker Provider Application

Reviewer: ```suggestion
## Create Sticker provider application
```

Author: done

> You can provide the sticker information to applications that use stickers. If you store the stickers into the Sticker database, the Sticker consumer applications can search for them.

Reviewer: ```suggestion
You can provide the sticker information to applications that use stickers. If you store the stickers in the Sticker database, the Sticker consumer applications can search for them.
```

Author: done

<a name="set_sticker_data"></a>

> ### Set Sticker Information to StickerData class

Author: done

> To store the sticker information to the Sticker database without using the JSON file, the sticker information must be set into the `Tizen.Uix.Sticker.StickerData` class:

Author: done

>     ```csharp

Reviewer: Hello @mijin85cho,
This seems very odd. The code tags on the markdown file have been declared properly. However, could you suggest any other reason for this issue?

>     ```csharp

Reviewer 3: Hello, @omar-srbd 
It's because of the indentation.

4 spaces indentations are needed when it belongs to a list item.
So line 90 is working well. 
(Line 90 belongs to line 88, the list item, "4. When the Sticker...").

However, even though line 106 is not a list item, there are 4 spaces indentation before line 108.
So the code block is broken.

Would you please remove the indentation in the case the code block is not under a list item?
Same thing is happening to line 143, etc.

>     ```csharp

Reviewer: Thank you for your detailed explanation of the issue.
Should I open another PR with the file `docs/application/dotnet/guides/uix/sticker.md` and implement changes to fix the code blocks?
Please let me know.

>     ```csharp

Reviewer 3: As the PR was merged, to update the content, yes, another PR is needed. Would you please open another PR to proceed it?

>     ```csharp

Reviewer: Sure, I will open the new PR right away.
Thank you.

>     ```csharp

Reviewer: @mijin85cho,
The changes on the code blocks have been implemented on https://github.com/Samsung/tizen-docs/pull/1720 as per your instruction.
Would you please check if everything is perfectly in place as it should be?

>     ```csharp

Reviewer 3: I checked that in PR 1720 the issue fixed.

![image](https://user-images.githubusercontent.com/9958374/198209393-569d785b-8ba2-4b78-8e05-cb0df6c3a7c5.png)

    StickerData SetStickerData()
    {
        try
        {
            /* Creates the StickerData class instance */
            StickerData stickerData = new StickerData();

            /* Sets the URI and URI type of the sticker */
            stickerData.SetUri(UriType.WebResource, "www.StickerProvider.com/emoji/smile");

            /* Adds a keyword of the sticker to the list */
            stickerData.AddKeyword("smile");
            stickerData.AddKeyword("cute");

            /* Sets the group name of the sticker */
            stickerData.GroupName = "Tizen";

            /* Sets the description of the sticker */
            stickerData.Description = "A yellow face with smiling eyes and full-toothed grin, as if saying Cheese! for the camera.";

            return stickerData;
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

<a name="set_json"></a>

> ### Set Sticker Information in JSON File

Author: done

> The following JSON file shows how to set the sticker information to be stored into the Sticker database:

Author: done

    {
        "sticker":
        [
            {
                "type" : 1,  /* Sets the URI type of the sticker */
                "uri" : "/res/face/slightly_smiling_face.png",  /* Sets the URI of the sticker */
                "keyword" : ["face", "smile"],  /* Adds a keyword of the sticker to the list */
                "group" : "face",  /* Sets the group name of the sticker */
                "thumbnail" : "/res/thumbnail/face/slightly_smiling_face.png",  /* Sets the thumbnail local path of the sticker */
                "description" : "A yellow face with simple, open eyes and a thin, closed smile."  /* Sets the description of the sticker */
            },
            {
                "type" : 1,
                "uri" : "/res/face/heart_eyes.png",
                "keyword" : ["face", "love", "adoration"],
                "group" : "face",
                "thumbnail" : "/res/thumbnail/face/heart_eyes.png",
                "description" : "A yellow face with an open smile, sometimes showing teeth, and red, cartoon-styled hearts for eyes."
            },
            ...
            {
                ...
            }
        ]
    }
    ```

<a name="insert_sticker"></a>

> ### Insert Sticker Information

Author: done

> You can use the following methods to store sticker information into the Sticker database:

Author: done

    ```csharp
    void InsertStickerData(StickerData stickerData)
    {
        try
        {
            StickerProvider.InsertData(stickerData);
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

- To store multiple sticker information, use `InsertData(string jsonPath)`:

    ```csharp
    async Task InsertStickerData(string jsonPath)
    {
        try
        {
            await StickerProvider.InsertData(jsonPath);
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

- To store sticker information without using `InsertData(StickerData data)` and `InsertData(string jsonPath)`, add the following metadata to the `tizen-manifest.xml` file:

  ```
  <category name="http://tizen.org/category/sticker"/>
  <!-- Sets the JSON local path -->
  <metadata key="http://tizen.org/metadata/sticker" value="/res/json/sticker.json"/>
  ```

  If you set the path of the JSON file to the value in metadata, sticker information of the JSON file is automatically stored when the Sticker provider application is installed.

  > **Note**

>   >

Author: done

  > The metadata must be added after category.

> ### Update Sticker Information

Author: done

You can update the sticker information stored in the Sticker database using the following method:

    ```csharp
    void UpdateStickerData(StickerData stickerData)
    {
        try
        {
            StickerProvider.UpdateData(stickerData);
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

<a name="consumer_application"></a>

> ## Create Sticker Consumer Application

Author: done

You can create the Sticker consumer application using `Tizen.Uix.Sticker.StickerConsumer` class to use sticker information. You can search for stickers and get details of stickers.

<a name="retrieve_sticker"></a>

> ### Retrieve Sticker Information

Author: done

You can retrieve all the stickers stored in the Sticker database and use keyword, group name, and type of URI to retrieve stickers.

The following example shows how to retrieve stickers using the group name:

    ```csharp
    void GetStickersByGroup(int offset, int count, string groupName)
    {
        try
        {
            IEnumerable<StickerData> stickers = StickerConsumer.GetStickersByGroup(offset, count, groupName);
            foreach(StickerData sticker in stickers)
            {
                /* Your code */
            }
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

You can retrieve the keywords and the group name of the stickers.

The following example shows how to retrieve the keyword list:

    ```csharp
    void GetKeywords()
    {
        try
        {
            IEnumerable<string> keywords = StickerConsumer.GetAllKeywords();
            foreach(string keyword in keywords)
            {
                /* Your code */
            }
        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

<a name="get_sticker_data"></a>

> ### Get Sticker Information

Author: done

        string groupName;
        string thumbnail;
        string description;
        DateTime updatedTime;
        DisplayType displayType;

        try
        {
            /* Gets the name of the sticker provider application */
            appId = stickerData.AppId;

            /* Gets the URI and URI type of the sticker */
            uri = stickerData.Uri;
            uriType = stickerData.UriType;

            /* Gets all the keywords of the sticker */
            keywords = stickerData.GetAllKeywords();

            /* Gets the group name of the sticker */
            groupName = stickerData.GroupName;

            /* Gets the thumbnail local path of the sticker */
            thumbnail = stickerData.Thumbnail;

            /* Gets the description of the sticker */
            description = stickerData.Description;

            /* Gets the last updated time of the sticker */
            updatedTime = stickerData.LastUpdated;

            /* Gets the display type of the sticker */
            displayType = stickerData.DisplayType;

            /* Your code */

        }
        catch (Exception e)
        {
            /* Error handling */
        }
    }
    ```

> ## Related Information

Reviewer: ```suggestion
## Related information
```

* Dependencies

>   - Tizen 7.0 and Higher for Mobile
>   - Tizen 7.0 and Higher for Wearable

Reviewer: ```suggestion
  - Tizen 7.0 and Higher
```

Author: done
